README.md
# IEEE754 conformance tests for Haskell

This project contains IEEE754R floating point conformance tests in Haskell.

There are two sets of tests so far:

##QTrial##
**QTrial** is the floating point benchmark described in *IEEE Standard 754 for Binary Floating-Point Arithmetic* by Prof. W. Kahan [Download link](http://www.eecs.berkeley.edu/~wkahan/ieee754status/IEEE754.PDF)

You should just be able to `cabal run qtrial` from the main directory or `runhaskell QTrial.hs` from `src`.

An extract from the result is:


```
Principal Tests:

Results for Float:
r = 4098.0 produces 12.0 and 12.0 sig. bits
r = 4098.25 fails: root 0.99989897 isn't at least 1
r = 4097.004 produces 12.0 and 11.999298 sig. bits
r =1.6777218e7 is too big for qtrail
r =1.6777218e7 is too big for qtrail
r =1.677722e7 is too big for qtrail
r =9.4906264e7 is too big for qtrail
r =9.4906264e7 is too big for qtrail
r =2.6843546e8 is too big for qtrail
r =2.6843546e8 is too big for qtrail
r =2.6843546e8 is too big for qtrail
r =2.6843546e8 is too big for qtrail
r =2.6843546e8 is too big for qtrail
r =4.2949673e9 is too big for qtrail
r =4.2949673e9 is too big for qtrail
Worst accuracy is 11.999298 sig. bits


Results for Double:
r = 4098.0 produces Infinity and Infinity sig. bits
r = 4098.25 produces Infinity and 53.0 sig. bits
r = 4097.00390625 produces Infinity and 53.451178091541244 sig. bits
r = 1.6777218e7 produces Infinity and Infinity sig. bits
r = 1.677721825e7 produces Infinity and 75.0 sig. bits
r = 1.6777219e7 produces Infinity and 71.0 sig. bits
r = 9.4906267e7 produces 26.499999994288153 and 26.499999986733027 sig. bits
r = 9.490626725e7 fails: root 0.999999995635551 isn't at least 1
r = 2.684354505e8 produces 28.0 and 27.999999919383132 sig. bits
r = 2.684354515e8 produces 28.0 and 27.99999993013205 sig. bits
r = 2.68435458e8 produces 28.0 and 28.0 sig. bits
r = 2.6843545825e8 produces 28.0 and 28.00000000268723 sig. bits
r = 2.6843545700000006e8 produces 28.0 and 27.999999989251084 sig. bits
r = 4.294967298e9 produces 32.0 and 32.0 sig. bits
r = 4.29496729825e9 produces 32.0 and 32.00000000016795 sig. bits
Worst accuracy is 26.499999986733027 sig. bits
```

This is fine for the Float cases, but the test illustrates an issue with the Double case
that is the subject of [this ticket](https://ghc.haskell.org/trac/ghc/ticket/9534).

##FPTest##

**FPTest** runs a sequence of floating point test vectors generated by
IBM's [FPGen Floating-Point Test Generator](https://www.research.ibm.com/haifa/projects/verification/fpgen/), which
is a commercial product.

This is done in two ways:

* By interpreting the tests
* By translating the tests into a Haskell HUnit test suite script that can just be run using `runhaskell`

The IBM test vectors are contained in the `test_suite` directory,
but you will want to get the latest version from the link above.

You will, most likely, want the binary tests unless you have a decimal machine.
Only 32-bit float test vectors are provided (not 64-bit double ones).

The syntax of the test vectors can be found [here](https://www.research.ibm.com/haifa/projects/verification/fpgen/syntax.txt) or in Chapter 4 of [Floating-Point Test-Suite for IEEE](https://www.research.ibm.com/haifa/projects/verification/fpgen/papers/ieee-test-suite-v2.pdf)

NB: The test vectors do not use the normal syntax for hexadecimal floating point literals.  The literals represent normal numbers
as '1.fPsigned exponent' and subnormal numbers by '0.fPsigned exponent'. The significand 'f' represents the 23-bit significand
for Floats and the 52-bit significand for Doubles.  Since a hexadecimal digit represents 4 bits, the question arises of where
the spare bit for Floats should sit.  In the standard representation (ie, set out in IEEE754 and implemented in the C 'dtoa'
and  'strtod' libraries, the spare 0 sits to the right (so the 23-bit significand 'f' is *left* justified in the 24-bit
hexadecimal representation).  The IBM representation *right* justifies the 23 bits of significand in the 24-bit hexadecimal
representation; the first hexadecimal digit of the significand represents 3 bits.
The IBM representation also has a fixed length format, including trailing zeros.

Some test vectors also uses lower case versions of certain specification elements.

### Results ###

The tests correctly identify some issues with GHC 7.8.3 that are being fixed or looked at:
* [signum and abs are incorrect](https://ghc.haskell.org/trac/ghc/ticket/7858)
* [the implementations of max / min are incorrect](https://ghc.haskell.org/trac/ghc/ticket/9530) 

### Caveats ###

A fair proportion of the test vector types is not implemented,
because there is no corresponding Haskell operation.  For example,
Haskell has no direct support for

* signalling NaNs
* the reporting of underflow or other exceptions
* rounding mode specification
* non-standard operations such as fused multiply add or operations that are applicable only to decimal floating point

There is not a great deal of point in testing some of these facilities from Haskell; where they are
accessible it would just be testing an underlying C library, which it might be better to
to directly.


## Other tests ##

* [FBench](https://www.fourmilab.ch/fbench/) by John Walker, founder of AutoDeck, Inc 
 is a complete optical design raytracing algorithm, shorn of its user interface.
 The benchmark, which has been going since 1980 has been ported from C to a variety of languages,
 including [Haskell](http://www.fourmilab.ch/fourmilog/archives/2012-09/001395.html)


changelog

0.2.2.0

	* Support a few more operations 	


test/HLint.hs
module Main (main) where

import Language.Haskell.HLint (hlint)
import System.Exit (exitFailure, exitSuccess)

arguments :: [String]
arguments =
    [
      "src"
    , "test"
    ]

main :: IO ()
main = do
    hints <- hlint arguments
    if null hints then exitSuccess else exitFailure


test/DocTest.hs
module Main (main) where

import System.FilePath.Glob (glob)
import Test.DocTest (doctest)

main :: IO ()
main = glob "src/*.hs" >>= doctest


test/Haddock.hs
module Main (main) where

import Data.List (genericLength)
import Data.Maybe (catMaybes)
import System.Exit (exitFailure, exitSuccess)
import System.Process (readProcess)
import Text.Regex (matchRegex, mkRegex)

average :: (Fractional a, Real b) => [b] -> a
average xs = realToFrac (sum xs) / genericLength xs

expected :: Fractional a => a
expected = 90

main :: IO ()
main = do
    output <- readProcess "cabal" ["haddock", "--executables"] ""
    if average (match output) >= expected
        then exitSuccess
        else putStr output >> exitFailure

match :: String -> [Int]
match = fmap read . concat . catMaybes . fmap (matchRegex pattern) . lines
  where
    pattern = mkRegex "^ *([0-9]*)% "


src/FPMain.hs
{- |
Module      : FPMain
Description : "FPMain" executes IBM floating point test cases either directly,
               or by generating a Haskell script that does it.
Copyright   : (c) John Pavel, 2014
                  IBM Corp 2004
License     : BSD3
Maintainer  : jrp@dial.pipex.com
Stability   : alpha
Portability : portable

See
<https://www.research.ibm.com/haifa/projects/verification/fpgen/papers/ieee-test-suite-v2.pdfi
Floating-Point Test-Suite for IEEE> and
<https://www.research.ibm.com/haifa/projects/verification/fpgen/syntax.txt
Syntax of the Test Cases>

-}
module Main where

import           FPRun

import           System.Console.GetOpt
import           System.Environment    (getArgs, getProgName)
import           System.Exit
import           System.IO

{-

TODO :: This initial version uses only basic Haskell, and so does not manipulate the floating point environment
for handling floating point exceptions, rounding mode, etc

-}


{- main = do
  {- quickCheck (prop_FloatHex :: Float -> Bool)
  quickCheck (prop_FloatHex :: Double -> Bool)
  evalTests -}
  translateTests
-}


data Options = Options {
    optAction       :: [FilePath] -> IO String,
    optOutputToFile :: String -> IO ()
  }

-- | Default command line option
defaultOptions :: Options
defaultOptions = Options {
    optAction = evalTestFiles,
    optOutputToFile = putStr
  }

-- | List of possible command line options
options :: [OptDescr (Options -> IO Options)]
options = [
  Option "e" ["evaluate"]
    (NoArg
      (\ opt -> return opt { optAction = evalTestFiles }))
    "Evaluate test script file (default)",

  Option "t" ["translate"]
    (NoArg
      (\ opt -> return opt { optAction = translateTestFiles }))
    "Translate test script file into a Haskell script",

  Option "o" ["output"]
    (ReqArg
      (\ arg opt -> return opt { optOutputToFile = writeFile arg })
      "FILE")
      "Output filename",

  Option "h" ["help"]
    (NoArg
      (\ _ -> do
        msg <- usage
        hPutStrLn stderr msg
        exitSuccess))
    "Show help"
  ]

-- | Return 'usage' string
usage :: IO String
usage = do
  prg <- getProgName
  return $ "Usage: " ++ usageInfo (prg ++ " [OPTIONS] SCRIPT...") options

{- parseArgs :: IO Options
parseArgs = do
argv <- getArgs
progName <- getProgName
let header = "Usage: " ++ progName ++ " [OPTION...] FILE1.fptest FILE2.fptest ..."
let helpMessage = usageInfo header (options "")
case getOpt RequireOrder (options helpMessage) argv of
(opts, [], []) -> foldlM (flip id) defaultOptions opts
(_, _, errs) -> ioError (userError (concat errs ++ helpMessage))
 -}

-- | 'main'
main :: IO ()
main = do
  {- quickCheck (prop_FloatHex :: Float -> Bool)
  quickCheck (prop_FloatHex :: Double -> Bool)
  evalTests -}

  -- Read the arguments
  args <- getArgs

  case getOpt Permute options args of

    (_, [], _) -> do
      u <- usage
      hPutStrLn stderr ("At least one test script required\n" ++ u)
      exitFailure

    (actions, scripts, []) ->
      do {- Should check that there is only one action,
            rather than running the last one given -}
        opts <- foldl (>>=) (return defaultOptions) actions

        -- Pull out the action and output action
        let Options { optAction = action,
        optOutputToFile = output } = opts

        {- feed the results of applying the action to the
           scripts to output -}
        action scripts >>= output

        exitSuccess

    (_, _, errs) -> do
      u <- usage
      hPutStrLn stderr (concat errs ++ u)

      exitFailure


testFiles :: [FilePath]
testFiles = map ("test_suite/" ++)
  [ "Basic-Types-Intermediate.fptest"]

saveTestFiles :: [FilePath]
saveTestFiles =
  ["Add-Cancellation-And-Subnorm-Result.fptest",
  "Add-Cancellation.fptest",
  "Add-Shift-And-Special-Significands.fptest",
  "Add-Shift.fptest",
  "Basic-Types-Inputs.fptest",
  "Basic-Types-Intermediate.fptest",
  "Compare-Different-Input-Field-Relations.fptest",
  "Corner-Rounding.fptest",
  "Decimal-Basic-Types-Inputs.fptest",
  "Decimal-Basic-Types-Intermediate.fptest",
  "Decimal-Clamping.fptest",
  "Decimal-Mul-Trailing-Zeros.fptest",
  "Decimal-Overflow.fptest",
  "Decimal-Rounding.fptest",
  "Decimal-Trailing-And-Leading-Zeros-Input.fptest",
  "Decimal-Trailing-And-Leading-Zeros-Result.fptest",
  "Decimal-Underflow.fptest",
  "Divide-Divide-By-Zero-Exception.fptest",
  "Divide-Trailing-Zeros.fptest",
  "Hamming-Distance.fptest",
  "Input-Special-Significand.fptest",
  "MultiplyAdd-Cancellation-And-Subnorm-Result.fptest",
  "MultiplyAdd-Cancellation.fptest",
  "MultiplyAdd-Shift-And-Special-Significands.fptest",
  "MultiplyAdd-Shift.fptest",
  "MultiplyAdd-Special-Events-Inexact.fptest",
  "MultiplyAdd-Special-Events-Overflow.fptest",
  "MultiplyAdd-Special-Events-Underflow.fptest",
  "Overflow.fptest",
  "Rounding.fptest",
  "Sticky-Bit-Calculation.fptest",
  "Underflow.fptest",
  "Rounding.fptest",
  "Sticky-Bit-Calculation.fptest",
  "Underflow.fptest",
  "Vicinity-Of-Rounding-Boundaries.fptest"]


src/QTrial.hs
{-|
Module      : Main
Description : Floating point benchmark described in /IEEE Standard 754 for Binary Floating-Point Arithmetic/ by Prof. W. Kahan
Copyright   : (c) John Pavel, 2014
                  Prof W Kahan, 1997
License     : BSD3
Maintainer  : jrp@dial.pipex.com
Stability   : experimental
Portability : portable

Adapted for Haskell from QTrial, described in /IEEE Standard 754 for Binary Floating-Point Arithmetic/
by
Prof. W. Kahan,
Elect. Eng. & Computer Science,
University of California,
Berkeley CA 94720-1776

<http://www.eecs.berkeley.edu/~wkahan/ieee754status/IEEE754.PDF>


This module runs the tests in Float, Double and CFloat and CDouble formats.

It would not be hard to adapt it to run other floating point formats.
-}

module Main where

import Data.Either
import Test.QuickCheck
import Foreign.C.Types
import Test.Framework (defaultMain, testGroup)
import Test.Framework.Providers.QuickCheck2

{- | 'qdrtc' computes the roots 'x1' and 'x2' of
quadradic of the form @p*x^2 - 2 q*x + r == 0@
as accuractely as they are determined by 'p', 'q', and 'r' -}
qdrtc :: RealFloat a => a -> a -> a -> (a, a)
qdrtc p q r
  | ss == 0 = (r / p, r / p)
  | otherwise = (r / ss, ss / p)
  where
    s = sqrt (q * q - p * r) -- NaN if sqrt (< 0)
    ss = q + copysign s q

-- Specialisations pf 'qdrtc' for different floating point formats
floatQdrtc :: Float -> Float -> Float -> (Float, Float)
floatQdrtc = qdrtc
doubleQdrtc :: Double -> Double -> Double -> (Double, Double)
doubleQdrtc = qdrtc
cfloatQdrtc :: CFloat -> CFloat -> CFloat -> (CFloat, CFloat)
cfloatQdrtc = qdrtc
cdoubleQdrtc :: CDouble -> CDouble -> CDouble -> (CDouble, CDouble)
cdoubleQdrtc = qdrtc


{- | @copysign(x, y)@ returns 'x' with the sign of 'y'.
Hence, @abs(x) = copysign( x, 1.0)@, even if 'x' is 'NaN'
NB: Haskell's 'signum' returns @signum NaN -> -1.0@ in
GHC 7.8.3, but should be fixed in 7.10 -}
copysign :: RealFloat a => a -> a -> a
copysign x y
  | isNaN y = 0 / 0
  | signum x == signum y = x
  | otherwise = -x

-- | 'log' base 2
log2 :: RealFloat a => a -> a
log2 x = logBase 2 (abs x)

{- | As Haskell's 'min' is not symmetric in its arguments in the
presence of a NaN, define a specific 'min'' -}
min' :: RealFloat a => a -> a -> a
min' x y
  | isNaN x || isNaN y = 0 / 0
  | otherwise = min x y

{- | 'qtrial' calculates the smaller root of @p*x^2 - 2 q*x + r == 0@, for
a given 'r', which determines 'p' and 'q' -}
qtrial :: (RealFloat a, Show a) => a -> Either String (a, String)
qtrial r
  | p <= 0 = Left "qtrail expects r > 2"
  | not ((r - q) == 1.0 && (q - p) == 1.0) = Left $ "r =" ++
      show r ++ " is too big for qtrail"
  | x1 < 1.0 = Left $ "r = " ++ show r ++ " fails: root " ++ show x1 ++
      " isn't at least 1"
  | otherwise = return (min' e1 e2, -- min' NaN NaN = NaN
                        "r = " ++ show r ++ " produces " ++ show e1 ++
                          " and " ++ show e2 ++ " sig. bits" )
  where
    p = r - 2.0
    q = r - 1.0

    (x1, x2) = qdrtc p q r

    e1 = - log2 (x1 - 1.0) -- Could be NaN
    e2 = - log2 ((x2 - 1.0) - 2.0 / p)

{- | 'qtestraw' runs the 'qtrial' against a series of different values of 'r'.
The later values will be too big for 32-bit arithmetic.
The result is either a 'String' reporting an error,
or a '(result, comment)' tuple -}
qtestraw :: (RealFloat a, Show a) => [Either String (a, String)]
qtestraw = map qtrial r
  where
  r =
    [2 ^ 12 + 2.0, -- for 24 sig. bits,
     2 ^ 12 + 2.25, -- and 6 hex.
     16 ^ 3 + 1.0 + 1.0 / 16 ^ 2, -- 6 hex. IBM.
     2 ^ 24 + 2.0, -- 48 bits CRAY -
     2 ^ 24 + 2.25, -- rounded;
     2 ^ 24 + 3.0, -- 48 bits chopped.
     94906267.0, -- 53 sig. bits.
     94906267 + 0.25, -- 53 sig. bits.
     2 ^ 28 - 5.5, -- PowerPC, i860.
     2 ^ 28 - 4.5, -- PowerPC, i860.
     2 ^ 28 + 2.0, -- 56 sig. bits,
     2 ^ 28 + 2.25, -- and 14 hex.
     16 ^ 7 + 1.0 + 1.0 / 16 ^ 6, -- 14 hex. IBM.
     2 ^ 32 + 2.0, -- 64 sig. bits.
     2 ^ 32 + 2.25] -- 64 sig. bits.

{- | 'qtest' generates a more readable rendering of the raw results of
produced by 'qtrial' -}
qtest :: (RealFloat a, Show a) => [Either String (a, String)] -> String
qtest raw = unlines (map (either id snd) raw) ++
    "Worst accuracy is " ++
    show (minimum (map fst (rights raw))) ++ " sig. bits"

-- Specialisations of 'qtestraw' for different floating point formats
floatQTest :: [Either String (Float, String)]
floatQTest = qtestraw
doubleQTest :: [Either String (Double, String)]
doubleQTest = qtestraw
cfloatQTest :: [Either String (CFloat, String)]
cfloatQTest = qtestraw
cdoubleQTest :: [Either String (CDouble, String)]
cdoubleQTest = qtestraw

{- 'QuickCheck' testing
Illustrates numerical instability, as need to ignore NaNs to pass -}

{- | '~==' is used for approximate equality, ignoring isNaN.
TODO:: overload for tuples, reduce tolerace to 1e-15 for doubles -}
infix 4 ~==
(~==) :: RealFloat a => a -> a -> Bool
x ~== y
  | isNaN x || isNaN y = True -- would otherwise generate false
  | otherwise = abs (x - y) < 1.0e-6 * abs x


-- The following test the invariant properties of 'qdrtc' given in the paper

test1 f p = p /= 0 ==> f p 0 0 == (0, 0)
prop_test1f = test1 floatQdrtc
prop_test1d = test1 doubleQdrtc
prop_test1cf = test1 cfloatQdrtc
prop_test1cd = test1 cdoubleQdrtc

test2 f = f 2.0 5.0 12.0 == (2, 3)
prop_test2f = test2 floatQdrtc
prop_test2d = test2 doubleQdrtc
prop_test2cf = test2 cfloatQdrtc
prop_test2cd = test2 cdoubleQdrtc

test3 f = f 2.0e-37 1.0 2.0 == (1.0, 1.0e37)
prop_test3f = test3 floatQdrtc
prop_test3d = test3 doubleQdrtc
prop_test3cf = test3 cfloatQdrtc
prop_test3cd = test3 cdoubleQdrtc

test4 f p q r =
  p /= 0 ==> x1 ~== 1.0 / x2' && x2 ~== 1.0 / x1'
  where
    (x1, x2) = f p q r
    (x1', x2') = f r q p
prop_test4f = test4 floatQdrtc
prop_test4d = test4 doubleQdrtc
prop_test4cf = test4 cfloatQdrtc
prop_test4cd = test4 cdoubleQdrtc

-- | TODO:: This test sometimes fails
test5 f mu p q r = (mu * p) /= 0 ==> x1 ~== x1' && x2 ~== x2'
  where
    (x1, x2) = f p q r
    (x1', x2') = f (mu * p) (mu * q) (mu * r)
prop_test5f = test5 floatQdrtc
prop_test5d = test5 doubleQdrtc
prop_test5cf = test5 cfloatQdrtc
prop_test5cd = test5 cdoubleQdrtc

{- (Orphan) instances, as they are defined neither in Foreign.C.Types,
not in Test.QuickCheck -}

instance Arbitrary CFloat where
  arbitrary = arbitrarySizedFractional
  shrink = shrinkRealFrac

instance Arbitrary CDouble where
  arbitrary = arbitrarySizedFractional
  shrink = shrinkRealFrac


-- Gather the QuickCheck unit tests
-- | 'unittests' checks that the implemention of 'qdrtc' satisfies
-- certain basic properties
unittests = [
    testGroup "test1" [
      testProperty "Float" prop_test1f,
      testProperty "Double" prop_test1d,
      testProperty "CFloat" prop_test1cf,
      testProperty "CDouble" prop_test1cd
    ],
    testGroup "test2" [
      testProperty "Float" prop_test2f,
      testProperty "Double" prop_test2d,
      testProperty "CFloat" prop_test2cf,
      testProperty "CDouble" prop_test2cd
    ],
    testGroup "test3" [
      testProperty "Float" prop_test3f,
      testProperty "Double" prop_test3d,
      testProperty "CFloat" prop_test3cf,
      testProperty "CDouble" prop_test3cd
    ],
    testGroup "test4" [
      testProperty "Float" prop_test4f,
      testProperty "Double" prop_test4d,
      testProperty "CFloat" prop_test4cf,
      testProperty "CDouble" prop_test4cd
    ],
    testGroup "test5" [
      testProperty "Float" prop_test5f,
      testProperty "Double" prop_test5d,
      testProperty "CFloat" prop_test5cf,
      testProperty "CDouble" prop_test5cd
    ]
  ]

-- | run the qtrial tests, followed by checking the properties of 'qdrtc'
main :: IO ()
main = do

  putStrLn "Principal Tests:"
  putStrLn "\nResults for Float:"
  putStrLn $ qtest floatQTest

  putStrLn "\nResults for CFloat:"
  putStrLn $ qtest cfloatQTest

  putStrLn "\nResults for Double:"
  putStrLn $ qtest doubleQTest

  putStrLn "\nResults for CDouble:"
  putStrLn $ qtest cdoubleQTest

  putStrLn "\n\nUnit Tests:"
  defaultMain unittests
